drivers/include/mfrc522.h

```diff
@@ -0,0 +1,810 @@
+ */
+void mfrc522_pcd_soft_power_down(mfrc522_t *dev);
+
+/**
+ * @brief Set the MFRC522 to leave soft power-down mode
+ *
+ * @param[in] dev  Device descriptor of the MFRC522
+ */
+void mfrc522_pcd_soft_power_up(mfrc522_t* dev);
+
+/**
+ * @brief Executes MFRC522_CMD_TRANSCEIVE.
+ *
+ * @note  CRC validation can only be done if back_data and back_len are specified
+ *
+ * @param[in]     dev         Device descriptor of the MFRC522
+ * @param[in]     send_data   Data to transfer to the FIFO
+ * @param[in]     send_len    Number of bytes to transfer to the FIFO
+ * @param[in]     back_data   Buffer if data should be read back after executing the command, otherwise NULL
+ * @param[in]     back_len    Max number of bytes to write to *back_data
+ * @param[out]    back_len    Number of bytes returned
+ * @param[in/out] valid_bits  Number of valid bits in the last byte. 0 for 8 valid bits
+ * @param[in]     rx_align    Defines the bit position in back_data[0] for the first bit received
+ * @param[in]     check_crc   If true, the last two bytes of the response are assumed to be a CRC_A that must be validated
+ *
+ * @return MFRC522_STATUS_OK on success, MFRC522_STATUS_??? otherwise
+ */
+mfrc522_status_code_t mfrc522_pcd_transceive_data(mfrc522_t *dev,
+                                                  uint8_t *send_data,
+                                                  uint8_t send_len,
+                                                  uint8_t *back_data,
+                                                  uint8_t *back_len,
+                                                  uint8_t *valid_bits,
+                                                  uint8_t rx_align,
+                                                  bool check_crc);
+
+/**
+ * @brief Transfers data to MFRC522's FIFO, executes a command, waits for
+ *        completion and transfers data back from the FIFO
+ *
+ * @note  CRC validation can only be done if back_data and back_len are specified
+ *
+ * @param[in]     dev         Device descriptor of the MFRC522
+ * @param[in]     command     The command to execute
+ * @param[in]     wait_irq    The bits in the ComIrqReg register that signals successful completion of the command
+ * @param[in]     send_data   Data to transfer to the FIFO
+ * @param[in]     send_len    Number of bytes to transfer to the FIFO
+ * @param[in]     back_data   Buffer if data should be read back after executing the command, otherwise NULL
+ * @param[in]     back_len    Max number of bytes to write to *back_data
+ * @param[out]    back_len    The number of bytes returned
+ * @param[in/out] valid_bits  The number of valid bits in the last byte. 0 for 8 valid bits.
+ * @param[in]     rx_align    Defines the bit position in back_data[0] for the first bit received
+ * @param[in]     check_crc   True => The last two bytes of the response is assumed to be a CRC_A that must be validated
+ *
+ * @return MFRC522_STATUS_OK on success, MFRC522_STATUS_??? otherwise
```

Comment:
I'm not sure whether I understand the doc correctly. Is there a `errno.h` where I should take these error codes from or should I declare them by myself but using the given naming scheme (and without an enum)? I doubt there is already a error code for a wrong crc value for example.

Reply:
I added error codes, namingly `-EINVAL` and `-ENOTSUP` to the `device_read` and `device_write` functions (they are just passed on from spi).
From the documentation: `MUST: all error and status return codes are named, e.g. DEVAB_OK, DEVAB_NOSPI, DEVAB_OVERFLOW, ...` I'd say `mfrc522_status_code_t` is following the rules. I really can't press status codes like `MFRC522_STATUS_COLLISION` or `MFRC522_STATUS_CRC_WRONG` into generic error codes. Especially because `MFRC522_STATUS_COLLISION` is used within the driver.